dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,42 +52,208 @@ public Pattern getParameterPattern() {
-                        columnName,
-                        Direction.INPUT,
-                        typeMapper.getJDBCSQLTypeForDBType(dataType),
-                        getJavaClass(dataType), position++);
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+                    allParams.put(NameNormaliser.normaliseName(paramName),
+                                  createColumnAccessor(paramName, rs));
+                }
             }
-            rs.close();
             return allParams;
         }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, columnDirection(rs));
+        }
+
+        protected Direction columnDirection(ResultSet rs) throws SQLException {
+            Direction columnDir = getColumnDirection(rs);
+            if (columnDir == null) {
+                throw new SQLException("Invalid column/parameter type");
+            }
+            return columnDir;
+        }
+
+        abstract protected Direction getColumnDirection(ResultSet rs) throws SQLException;
+
+        protected DbParameterAccessor createDbParamAccessor(String paramName, ResultSet rs, Direction direction)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(paramName, direction,
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+    }
+
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getColumnDirection(ResultSet rs) {
+            return Direction.INPUT;
+        }
+    }
+
+    abstract private class DatabaseExecutable extends DatabaseObject {
+
+        private DatabaseExecutable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected Direction getColumnDirection(ResultSet rs) throws SQLException {
+            return columnTypeToDirection(rs.getInt("COLUMN_TYPE"));
+        }
+
+        abstract protected Direction columnTypeToDirection(int columnType);
```

Comment:
This trade off consideration is of course another subjective area. 2 levels is more obscure (irrelevant concepts for some cases) whereas 3 levels is more complex but is more expressive. I think this is something where you are probably more experienced than me (certainly in the Java space) so I don't mind which way we go.